design/design-generic-references.md

```diff
@@ -0,0 +1,217 @@
+# Generic Resource References
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Accepted
+
+## Background
+
+In many cases, Crossplane managed resources require information that is not
+available when the user creates the resource. One example is that if you'd like to create
+a `Subnet`, you need to supply a VPC ID but if you create the `VPC` at the same
+time then the ID will be available only after the creation, so you need to extract
+the ID and use it in `Subnet`. Today, it's possible to let the provider do this
+operation using cross-resource references. However, this requires knowing the kind of
+the referenced resource and how to extract the value during development time of
+the provider. There are several cases where this is not possible:
+
+* The referencer field may not always target a single kind. For example, a URL
+  field in Route53 could get populated by a URL of an S3 bucket or access endpoint
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
+In order to enable such use cases, we need a generic version of the cross-resource
+references where you can specify all metadata required to find the referenced
+resource as well as details about how to extract the value you need and where to
+put it in the referencer object.
+
+## Proposal
+
+All managed resources will have a top-level spec field that lists the generic
+references. They will be resolved in order and failure of any will cause
+reconciliation to fail. Inspired from the [latest developments in API conventions](https://github.com/kubernetes/community/pull/5748)
+regarding multi-kind references, the API will look like the following:
+```yaml
+# A generic reference to another managed resource.
+spec:
+  patches:
+  - fromObject:
+      apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      name: main-vpc # optional, matchLabels can be used instead.
+      fieldPath: spec.forProvider.cidrBlock
+    toFieldPath: spec.cidrBlock
+```
+```yaml
+# A generic selector to another managed resource.
+spec:
+  patches:
```

Comment:
Regarding the naming, I saw there was a discussion in this PR before. I am curious whether it could be made a bit more generic so this design can cover https://github.com/crossplane/crossplane/issues/2072#issuecomment-764001935 as well.

In that scenario, instead of patching, what people want is to defer the managed resource handing till its dependent resources are resolvable. I like the idea that @negz suggested in the above comment for https://github.com/crossplane/crossplane/issues/2072 since it can use the same solution to address two different scenarios, patching vs. dependency check.

With that, I guess using patching in the above scenario would be a bit misleading. May be `depends` is better? Also, both `fieldPath` and `toFieldPath` could be optional. When `toFieldPath` is omitted, it does not patch anything. Instead, it will try to resolve the dependent resource(s) when `fieldPath` is omitted, or the `fieldPath` value(s) when `fieldPath` is specified.

Does it make sense?

@muvaf

Reply:
The naming depends on the main thing that the mechanism is trying to achieve, right? In that sense, its main goal is to be able to gather information from other sources and put it as value to a field, i.e. patching. It provides a dependency graph as well because it's blocking but I'm not sure that's a big enough goal of this design that it should affect what it's called.

Reply:
> both fieldPath and toFieldPath could be optional. When toFieldPath is omitted, it does not patch anything. Instead, it will try to resolve the dependent resource(s) when fieldPath is omitted, or the fieldPath value(s) when fieldPath is specified.

You can achieve this by getting `metadata.uid` of the source and patching an arbitrary label/annotation of the resource. I'd like to look for more feedback from the community to make it a first citizen use case that we should account for.

Reply:
> I'm not sure that's a big enough goal of this design that it should affect what it's called.
> I'd like to look for more feedback from the community to make it a first citizen use case that we should account for.

Thanks @muvaf, yeah, I can understand your point. The thing is, if we decide to leave this door open to wait for feedback, and when we see more requests on this, we may have to deprecate something and use another more appropriate term.

On the other hand, if we use a more generic term here, `dependencies`, `references` or something better, even we do not want to support the "creation deferral" case at the moment, this will give us a bit more flexible to make turnaround.

> You can achieve this by getting metadata.uid of the source and patching an arbitrary label/annotation of the resource.
I can see that works. It's more like a workaround for now :-D

BTW: I was trying to experiment "creation deferral" using provider-kubernetes by following the spec defined in this proposal, hopefully I can share more on this later based on my findings.